src/Symfony/Component/HttpClient/RetryHttpClient.php

```diff
@@ -0,0 +1,100 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\HttpClient;
+
+use Psr\Log\LoggerInterface;
+use Psr\Log\NullLogger;
+use Symfony\Component\HttpClient\Response\AsyncContext;
+use Symfony\Component\HttpClient\Response\AsyncResponse;
+use Symfony\Component\HttpClient\Retry\RetryBackOffInterface;
+use Symfony\Component\HttpClient\Retry\RetryDeciderInterface;
+use Symfony\Contracts\HttpClient\ChunkInterface;
+use Symfony\Contracts\HttpClient\Exception\TransportExceptionInterface;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+use Symfony\Contracts\HttpClient\ResponseInterface;
+
+/**
+ * Automatically retries the failing requests.
+ *
+ * @author Jérémy Derussé <jeremy@derusse.com>
+ */
+class RetryHttpClient implements HttpClientInterface
+{
+    use AsyncDecoratorTrait;
+
+    private $decider;
+    private $strategy;
+    private $maxRetries;
+    private $logger;
+
+    private $retryCount = 0;
+
+    /**
+     * @param int $maxRetries The maximum number of times to retry
+     */
+    public function __construct(HttpClientInterface $client, RetryDeciderInterface $decider, RetryBackOffInterface $strategy, int $maxRetries = 3, LoggerInterface $logger = null)
+    {
+        $this->client = $client;
+        $this->decider = $decider;
+        $this->strategy = $strategy;
+        $this->maxRetries = $maxRetries;
+        $this->logger = $logger ?: new NullLogger();
+    }
+
+    public function request(string $method, string $url, array $options = []): ResponseInterface
+    {
+        return new AsyncResponse($this->client, $method, $url, $options, function (ChunkInterface $chunk, AsyncContext $context) use ($method, $url, $options) {
+            $exception = null;
+            try {
+                // clone the chunk to let the original value of `$didthrow`
+                $clonedChunk = clone $chunk;
+                if ($clonedChunk->isTimeout() || null !== $clonedChunk->getInformationalStatus()) {
+                    yield $chunk;
+
+                    return;
+                }
+
+                // only retry first chunk
+                if (!$clonedChunk->isFirst()) {
+                    $context->passthru();
+                    yield $chunk;
+
+                    return;
+                }
+            } catch (TransportExceptionInterface $exception) {
+                // catch TransportExceptionInterface to send it to strategy.
+            }
+
+            if ($this->retryCount >= $this->maxRetries || !$this->decider->shouldRetry($method, $url, $options, $response = $context->getResponse(), $exception)) {
```

Comment:
we should not pass the original response to the deciders: reading any method of it will disable the failsafe destructor.
Instead, one should read all the needed state via the dedicated `$context->get*` methods.
If we want to pass a ReponseInterface to the decider+backoff, we could wrap these values in a `MockResponse`. Or we could pass the headers+status code as raw data.

Reply:
sometime, content could help to decide if the response should be retry (ie. deserialize the response to get details about the error. This is what is done in official AWS SDK for instance to handle `ThrottlingException`, or `ProvisionedThroughputExceededException`).

From another side, I don't want to ALWAYS read headers/content... In most cases, status code is enough

Any idea how to read responses on demand, without breaking the "failsafe" state?

Reply:
hmmm.. issue with "decider need access to the response" is, it break the stream. We can't both stream the response (let say a several MB response) and at the same time, read the entire response to define if the response should be retried...

I change the contract to take in parameter a `StatusCode` and a `Headers`. If one day somebody want the response body, we'll open a PR.

Reply:
CHeck https://github.com/symfony/symfony/blob/9c8cd0827f6fab9bf840b7fd2940bcbe1b16e75e/src/Symfony/Component/HttpClient/Response/AsyncResponse.php#L335-L343